### PR TITLE
[RFE#1703] feat: add API to insert menu item

### PR DESCRIPTION
**Pull request by miurahr**
Resolved RFE#1703

This introduces a menu manipulation utility `MenuExternder` class.

There are `get(Something)Menu` methods which returns JMenu object in `IMainMenu` interface.
It is not stable to insert menu item from plugins when multiple plugins insert menu items
or new menu item inserted for new OmegaT versions.

It add an API `IMainMenu#getMenu(MenuMarker)`. It is designed to extensible.
When OmegaT add a new `JMenu` on the menubar, we don't need to extend the interface.
Otherwise, we will only update a `MenuKey` enum class.

We have now `MenuExtender#addMenuItem` method to allow plugins/modules inserts its menu item safely.
The method takes `MenuKey` enum value to indicate where to add menu item.

## Pull request type

- Feature enhancement -> [enhancement]

## Which ticket is resolved?

- Request to add new API to allow plugins/modules to add menu item in requested place
     *  https://sourceforge.net/p/omegat/feature-requests/1703

## What does this PR change?

- Update IMainMenu to add new API to get menu
- Use MenuKey enum class for operations.
- Introduce menu manipulation util class
- Add test cases
- MenuItemPager is also aware of Extender API
- style: apply spotless

### CODE REVIEW

**Comment by t-cordonnier** [comment]
Except the comments in the code, this is globally a good idea but
1. an example should be put somewhere, to be sure the meaning of all parameters is clear
2. Are you sure it works if I want to add two menus, one after the other, from the same plugin? Please test with two menus and prepare an example to be sure we don't need to add them in the reverse order

**Comment by miurahr** [comment]
> Except the comments in the code, this is globally a good idea but
> 
> 1. an example should be put somewhere, to be sure the meaning of all parameters is clear
> 2. Are you sure it works if I want to add two menus, one after the other, from the same plugin? Please test with two menus and prepare an example to be sure we don't need to add them in the reverse order

This PR is originaly a work in PR #592 which add a menu item  in help menu.
A feature has been improved through discussion on issue ticket and here.

I agree that it is better to have a 1) test case and 2) actual use case.

I'd like to try migrate  MED feature and external finder for use case.

**Comment by t-cordonnier** [comment]
Please wait before touching MED feature.
I am currently working on the possibility to convert the two formats OMT and MED to a new plugin type named "packages" (which could later be used to implement other package formats, such as the ones used by other CAT tools)
In this new configuration, all formats use the same menus but the difference is made in the open/save window.
For this reason I ask you not to use this example. Sorry. But external-finder is probably a better sample.

File: src/org/omegat/externalfinder/ExternalFinder.java [hunk 1]
```diff
@@ -121,15 +122,13 @@ private void onLoad() {
...
-
                 // Separator
                 Component separator = new JPopupMenu.Separator();
-                toolsMenu.add(separator);
+                Core.getMainWindow().addMenuItem(MenuKey.TOOLS, separator);
                 menuItems.add(separator);
 
                 // add menuItems to menu
-                MenuItemPager pager = new MenuItemPager(toolsMenu);
+                MenuItemPager pager = new MenuItemPager(Core.getMainWindow().getMainMenu().getToolsMenu());
```
Comment (miurahr) [inline src/org/omegat/externalfinder/ExternalFinder.java:131]: TODO: handle pager

File: src/org/omegat/gui/main/MenuKey.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+package org.omegat.gui.main;
+
+public enum MenuKey implements MenuMarker {
+    PROJECT("project", 14),
+    EDIT("edit", 15),
+    GOTO("goto", 18),
+    VIEW("view", 32),
+    TOOLS("tools", 5),
+    OPTIONS("options", 13),
+    HELP("help", 2),
```
Comment (miurahr) [inline src/org/omegat/gui/main/MenuKey.java:35]: TODO: should test or check these numbers.

File: src/org/omegat/gui/main/MainWindow.java [hunk 3]
```diff
@@ -228,10 +232,26 @@ public Font getApplicationFont() {
...
+    @Override
+    public void addMenuItem(MenuKey target, Component menuItem) {
+        if (target == null) {
+            throw new NullPointerException();
+        }
+        if (menuItem == null) {
+            throw new NullPointerException();
+        }
+        JMenu targetMenu = menu.getMenu(target);
+        targetMenu.add(menuItem, target.getPosition());
```
Comment (miurahr) [inline src/org/omegat/gui/main/MainWindow.java:252]: when multiple menu added, second item will be put above first item.

**Review by miurahr — COMMENTED** [review]
Need further improvement

**Comment by miurahr** [comment]
I added the test.

- testAddGotoMenuItem checks a case when extension point is end of menu items (-1)
- testAddHelpMenuItem checks a case when extension point is middle of menu items.
- testAddOptionsMenuItem checks a case when extension point is middle of menu items and some menu item is JMenu instance.

Each case try to reproduce actual menu items in TestCore.java for test.

**Comment by miurahr** [comment]
feature added:  add items with pager and test it.

TODO:  removal of added items with pager and its test case.

**Comment by miurahr** [comment]
Dear reviewer, @t-cordonnier,  could you look a test case and an example in external finder feature?

There are several cases 1) simple addition, 2) multiple addition, 3) addition to middle of menu list, 4)using pager.
I've updated TestCore.java to set menu tree as a test data.

**Comment by miurahr** [comment]
I've satisfied for the changes now.

**Comment by miurahr** [comment]
After self-reviewing my changes, I think we don't need to change IMainWindow interface.
Here we are doing only addtion of `IMainMenu#getMenu` method.

Now `MenuKey` is an internal enum of `MenuExtender`.

File: src/org/omegat/gui/main/BaseMainWindowMenu.java [hunk 4]
```diff
@@ -566,6 +581,8 @@ protected void constructMenu() {
         toolsMenu.add(toolsShowStatisticsMatchesPerFileMenuItem);
         toolsMenu.addSeparator();
         toolsMenu.add(toolsAlignFilesMenuItem);
+        // extension point
+        toolsMenu.addSeparator();
```
Comment (miurahr) [inline src/org/omegat/gui/main/BaseMainWindowMenu.java:585]: It is better not to change menu itself.

File: src/org/omegat/gui/main/BaseMainWindowMenu.java [hunk 5]
```diff
@@ -537,6 +548,8 @@ protected void constructMenu() {
         gotoMenu.addSeparator();
         gotoMenu.add(gotoNotesPanelMenuItem);
         gotoMenu.add(gotoEditorPanelMenuItem);
+        // extension point
+        gotoMenu.addSeparator();
```
Comment (miurahr) [inline src/org/omegat/gui/main/BaseMainWindowMenu.java:552]: same here

File: src/org/omegat/gui/main/BaseMainWindowMenu.java [hunk 6]
```diff
@@ -556,6 +569,8 @@ protected void constructMenu() {
         viewModificationInfoMenu.add(viewDisplayModificationInfoNoneRadioButtonMenuItem);
         viewModificationInfoMenu.add(viewDisplayModificationInfoSelectedRadioButtonMenuItem);
         viewModificationInfoMenu.add(viewDisplayModificationInfoAllRadioButtonMenuItem);
+        // extension point
+        viewMenu.addSeparator();
```
Comment (miurahr) [inline src/org/omegat/gui/main/BaseMainWindowMenu.java:573]: same here

File: src/org/omegat/gui/main/BaseMainWindowMenu.java [hunk 7]
```diff
@@ -594,12 +611,14 @@ protected void constructMenu() {
         optionsMenu.add(optionsSetupFileFiltersMenuItem);
         optionsMenu.add(optionsSentsegMenuItem);
         optionsMenu.add(optionsWorkflowMenuItem);
+        // extension point
         optionsMenu.addSeparator();
-        optionsMenu.add(optionsAccessConfigDirMenuItem);
         optionsMenu.addSeparator();
+        optionsMenu.add(optionsAccessConfigDirMenuItem);
```
Comment (miurahr) [inline src/org/omegat/gui/main/BaseMainWindowMenu.java:617]: it is better not to change order

File: src/org/omegat/gui/main/ConsoleWindow.java [hunk 8]
```diff
@@ -138,10 +138,12 @@ public IMainMenu getMainMenu() {
         throw new NoSuchMethodError("Invalid call of ConsoleWindow");
     }
 
+    @Override
```
Comment (miurahr) [inline src/org/omegat/gui/main/ConsoleWindow.java:141]: This is unrelated change.

File: src/org/omegat/gui/main/MainWindow.java [hunk 9]
```diff
@@ -182,8 +183,10 @@ public void windowDeactivated(WindowEvent we) {
         CoreEvents.registerApplicationEventListener(new IApplicationEventListener() {
             public void onApplicationStartup() {
                 MainWindowUI.initializeScreenLayout(MainWindow.this);
-
-                UIDesignManager.removeUnusedMenuSeparators(menu.getOptionsMenu().getPopupMenu());
+                // Remove all duplicated separators from all menus
+                for (final MenuExtender.MenuKey key : MenuExtender.MenuKey.values()) {
+                    UIDesignManager.removeUnusedMenuSeparators(menu.getMenu(key).getPopupMenu());
+                }
```
Comment (miurahr) [inline src/org/omegat/gui/main/MainWindow.java:189]: It should be another PR and review.

File: src/org/omegat/util/gui/MenuExtender.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+         */
+        OPTIONS("options", Platform.isMacOSX() ? 9 : 11),
+        /**
+         * Help menu.
+         */
+        HELP("help", 3),
+        /**
+         * Open recent pop-up menu in Project menu.
+         */
+        PROJECTOPENRECENT("projectOpenRecent", 0),
```
Comment (miurahr) [inline src/org/omegat/util/gui/MenuExtender.java:73]: Not all the JMenu object is not a target of extension.

File: src/org/omegat/gui/main/BaseMainWindowMenu.java [hunk 11]
```diff
@@ -730,10 +738,15 @@ public void invokeAction(String action, int modifiers) {
      *            title name key in resource bundle
      * @return menu instance
      */
-    protected JMenu createMenu(final String titleKey) {
+    protected JMenu createMenu(String titleKey) {
+        return createMenu(null, titleKey);
+    }
+
+    protected JMenu createMenu(MenuExtender.MenuKey menuKey, String titleKey) {
```
Comment (miurahr) [inline src/org/omegat/gui/main/BaseMainWindowMenu.java:745]: nullable parameter should be last, and should check nullity.

File: src/org/omegat/gui/main/BaseMainWindowMenu.java [hunk 12]
```diff
@@ -436,6 +441,8 @@ protected void constructMenu() {
         projectMenu.addSeparator();
         projectMenu.add(projectMedOpenMenuItem);
         projectMenu.add(projectMedCreateMenuItem);
+        // extension point
+        projectMenu.addSeparator();
```
Comment (miurahr) [inline src/org/omegat/gui/main/BaseMainWindowMenu.java:445]: Don't change menu itself.

File: src/org/omegat/gui/main/BaseMainWindowMenu.java [hunk 13]
```diff
@@ -508,6 +515,7 @@ protected void constructMenu() {
         editMenu.add(editRegisterUntranslatedMenuItem);
         editMenu.add(editRegisterEmptyMenuItem);
         editMenu.add(editRegisterIdenticalMenuItem);
+        editMenu.addSeparator();
```
Comment (miurahr) [inline src/org/omegat/gui/main/BaseMainWindowMenu.java:518]: don't change menu itself.

**Review by miurahr — COMMENTED** [review]
Change should be minimum for single PR.

**Comment by miurahr** [comment]
Self-review and all ajustments are done.
Now ready for merge.

File: test/fixtures/org/omegat/core/TestCore.java [hunk 14]
```diff
@@ -72,33 +75,145 @@ public final void setUpCore() throws Exception {
...
+                    helpMenu.add(new JMenuItem("About"));
+                    helpMenu.addSeparator();
+                    helpMenu.add(new JMenuItem("item 3"));
+                }
+                return helpMenu;
+            }
+
+            @Override
+            public JMenu getMenu(final MenuExtender.MenuKey marker) {
+                switch (marker) {
```
Comment (miurahr) [inline test/fixtures/org/omegat/core/TestCore.java:182]: Case for project menu is missing.

**Comment by miurahr** [comment]
Further fixes are done.

**Comment by brandelune** [comment]
Please:
- add a line to changes.txt

**Comment by miurahr** [comment]
> Please:
> 
> * add a line to changes.txt

Done. I'd like to squash and merge.
